### PR TITLE
Payment Channels (RIPD-1224)

### PR DESCRIPTION
**Pull request by seelabs**
This PR add unidirectional payment channels.

Payment channels permit off-ledger checkpoints of XRP payments flowing in a single direction. A channel sequesters the owner's XRP in its own ledger entry. The owner can authorize the recipient to claim up to a given balance by giving the receiver a signed message (off-ledger). The recipient can use this signed message to claim any unpaid balance while the channel remains open. The owner can top off the line as needed. If the channel has not paid out all its funds, the owner must wait out a delay to close the channel to give the recipient a chance to supply any claims. The recipient can close the channel at any time. Any transaction that touches the channel after the expiration time will close the channel. The total amount paid increases monotonically as newer claims are issued. When the channel is closed any remaining balance is returned to the owner. Channels are intended to permit intermittent off-ledger settlement of ILP trust lines as balances get substantial. For bidirectional channels, a payment channel can be used in each direction.

There are three new transactions (See the `PayChan.cpp` file for documentation on these Txs):
* PaymentChannelCreate
* PaymentChannelFund
* PaymentChannelClaim

There are three new RPC commands:
* account_channels queries the ledger for channels (much like account_lines)
* channel_authorize signs authorizations
* channel_verify verifying authorizations

### CODE REVIEW

File: src/ripple/app/tx/impl/PayChan.cpp [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+            if (ctx_.tx[sfPublicKey] != pk)
+                return temBAD_SIGNER;
+            Serializer msg;
+            std::uint8_t const hdr[4] = {'C', 'L', 'M', 0};
+            msg.addRaw (hdr, 4);
+            msg.add256 (k.key);
+            msg.add64 (authAmt.drops ());
+            if (!verify (pk, msg.slice (), *sig, /*canonical*/ true))
+                return temBAD_SIGNATURE;
+        }
```
Comment (JoelKatz) [inline src/ripple/app/tx/impl/PayChan.cpp]: You cannot check the signature in `doApply`. You must do it in `preclaim`. For the most obvious reason why it must be this way, consider what happens if we later have to retry the transaction against a different ledger. We don't want to have to check the signature again.

Comment (seelabs) [inline src/ripple/app/tx/impl/PayChan.cpp]: I moved the signature check into `preflight`. You suggested `preclaim`, but since the tx has the public key it can be checked without a ledger (the public key on the tx _is_ checked against the one stored in the ledger later.)

File: src/ripple/rpc/handlers/AccountChannels.cpp [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    AccountID raDstAccount;
+    if (hasDst)
+    {
+        result = RPC::accountFromString (raDstAccount, strDst);
+        if (result)
+            return result;
+    }
+
+    unsigned int limit;
+    if (auto err = readLimitField(limit, RPC::Tuning::accountLines, context))
```
Comment (miguelportilla) [inline src/ripple/rpc/handlers/AccountChannels.cpp]: Use RPC::Tuning::accountChannels

File: src/ripple/app/tests/PayChan_test.cpp [hunk 3]
```diff
@@ -0,0 +1,689 @@
...
+        auto const bob = Account ("bob");
+        env.fund (XRP (10000), alice, bob);
+        auto const pk = alice.pk ();
+        auto const settleDelay = 100s;
+        env (create (alice, bob, XRP (1000), settleDelay, pk));
+        auto const chan = channel (*env.current (), alice, bob);
+        expect (channelBalance (*env.current (), chan) == XRP (0));
+        expect (channelAmount (*env.current (), chan) == XRP (1000));
+
+        env (fund (alice, chan, XRP (1000)));
```
Comment (miguelportilla) [inline src/ripple/app/tests/PayChan_test.cpp]: Consider testing a non XPR with fund (should be temBAD_AMOUNT?).

File: src/ripple/app/tests/PayChan_test.cpp [hunk 4]
```diff
@@ -0,0 +1,689 @@
...
+        testcase ("simple");
+        using namespace jtx;
+        using namespace std::literals::chrono_literals;
+        Env env (*this, features (featurePayChan));
+        auto const alice = Account ("alice");
+        auto const bob = Account ("bob");
+        env.fund (XRP (10000), alice, bob);
+        auto const pk = alice.pk ();
+        auto const settleDelay = 100s;
+        env (create (alice, bob, XRP (1000), settleDelay, pk));
```
Comment (miguelportilla) [inline src/ripple/app/tests/PayChan_test.cpp]: Consider testing a non XPR with fund (should be temBAD_AMOUNT?).

File: src/ripple/app/tests/PayChan_test.cpp [hunk 5]
```diff
@@ -0,0 +1,689 @@
...
+        expect (chanAmt == XRP (2000));
+
+        {
+            // No signature needed since the owner is claiming
+            auto const preBob = env.balance (bob);
+            auto const delta = XRP (500);
+            auto const reqBal = chanBal + delta;
+            auto const authAmt = reqBal + XRP (100);
+            assert (reqBal <= chanAmt);
+            env (claim (alice, chan, reqBal, authAmt));
```
Comment (miguelportilla) [inline src/ripple/app/tests/PayChan_test.cpp]: Since the TransactionType is different from create and fund, consider testing non XPR with claim.

File: src/ripple/app/tests/PayChan_test.cpp [hunk 6]
```diff
@@ -0,0 +1,689 @@
...
+            expect (channelAmount (*env.current (), chan) == chanAmt);
+            expect (env.balance (bob) == preBob + delta);
+            chanBal = reqBal;
+        }
+        {
+            auto preBob = env.balance (bob);
+            auto const delta = XRP (500);
+            auto const reqBal = chanBal + delta;
+            auto const authAmt = reqBal + XRP (100);
+            assert (reqBal <= chanAmt && reqBal <= authAmt);
```
Comment (miguelportilla) [inline src/ripple/app/tests/PayChan_test.cpp]: 'reqBal <= authAmt' ?

File: src/ripple/app/tests/PayChan_test.cpp (hunk 5 above)
Comment (miguelportilla) [inline src/ripple/app/tests/PayChan_test.cpp]: Attempt a no signature claim for more than authorized? I realize a signed claim for more than authorized is tested later on.

Comment (miguelportilla) [inline src/ripple/app/tests/PayChan_test.cpp]: Also consider testing negative XRP values with the create, fund and claim transactions.

File: src/ripple/app/tests/PayChan_test.cpp [hunk 7]
```diff
@@ -0,0 +1,752 @@
...
+            return XRPAmount{-1};
+        return (*slep)[sfAmount];
+    }
+
+    static
+    boost::optional<std::int64_t>
+    channelExpiration (ReadView const& view, uint256 const& chan)
+    {
+        Keylet const k (ltPAYCHAN, chan);
+        auto const slep = view.read (k);
```
Comment (vinniefalco) [inline src/ripple/app/tests/PayChan_test.cpp]: Consider
```
auto const sle = view.read({ltPAYCHAN, id});
```

File: src/ripple/app/tests/PayChan_test.cpp [hunk 8]
```diff
@@ -0,0 +1,689 @@
...
+            STAmount const reqAmt = authAmt + 1;
+            assert (reqAmt <= chanAmt);
+            auto const sig =
+                signClaimAuth (alice.pk (), alice.sk (), chan, authAmt);
+            env (claim (bob, chan, reqAmt, authAmt, Slice (sig), alice.pk ()),
+                ter (tecNO_PERMISSION));
+            expect (channelBalance (*env.current (), chan) == chanBal);
+            expect (channelAmount (*env.current (), chan) == chanAmt);
+            auto const feeDrops = env.current ()->fees ().base;
+            expect (env.balance (bob) == preBob + -feeDrops);
```
Comment (miguelportilla) [inline src/ripple/app/tests/PayChan_test.cpp]: Should be 'preBob - feeDrops'

File: src/ripple/app/tests/PayChan_test.cpp [hunk 9]
```diff
@@ -0,0 +1,689 @@
...
+            auto const reqBal = chanBal + delta;
+            auto const authAmt = reqBal + XRP (100);
+            assert (reqBal <= chanAmt && reqBal <= authAmt);
+            auto const sig =
+                signClaimAuth (alice.pk (), alice.sk (), chan, authAmt);
+            env (claim (bob, chan, reqBal, authAmt, Slice (sig), alice.pk ()));
+            expect (!channelExists (*env.current (), chan));
+            auto const feeDrops = env.current ()->fees ().base;
+            expect (env.balance (alice) == preAlice + chanAmt - chanBal);
+            expect (env.balance (bob) == preBob + - feeDrops);
```
Comment (miguelportilla) [inline src/ripple/app/tests/PayChan_test.cpp]: Should be 'preBob - feeDrops'

File: src/ripple/app/tests/PayChan_test.cpp [hunk 10]
```diff
@@ -0,0 +1,689 @@
...
+            auto const authAmt = reqBal + XRP (100);
+            assert (reqBal <= chanAmt);
+            env (claim (alice, chan, reqBal, authAmt));
+            expect (channelBalance (*env.current (), chan) == reqBal);
+            expect (channelAmount (*env.current (), chan) == chanAmt);
+            expect (env.balance (bob) == preBob + delta);
+            chanBal = reqBal;
+        }
+        {
+            auto preBob = env.balance (bob);
```
Comment (miguelportilla) [inline src/ripple/app/tests/PayChan_test.cpp]: Suggest comment: // claim with signature

File: src/ripple/app/tests/PayChan_test.cpp [hunk 11]
```diff
@@ -0,0 +1,689 @@
...
+        env.fund (XRP (10000), alice, bob);
+        auto const pk = alice.pk ();
+        auto const settleDelay = 100s;
+        env (create (alice, bob, XRP (1000), settleDelay, pk));
+        auto const chan = channel (*env.current (), alice, bob);
+        expect (channelBalance (*env.current (), chan) == XRP (0));
+        expect (channelAmount (*env.current (), chan) == XRP (1000));
+
+        env (fund (alice, chan, XRP (1000)));
+        auto chanBal = channelBalance (*env.current (), chan);
```
Comment (miguelportilla) [inline src/ripple/app/tests/PayChan_test.cpp]: Should probably check alice's balance here, just to verify all is peachy.

File: src/ripple/app/tests/PayChan_test.cpp [hunk 12]
```diff
@@ -0,0 +1,689 @@
...
+        auto chanAmt = channelAmount (*env.current (), chan);
+        expect (chanBal == XRP (0));
```
Comment (miguelportilla) [inline src/ripple/app/tests/PayChan_test.cpp]: Should probably check alice's balance here, just to verify all is peachy.

File: src/ripple/app/tests/PayChan_test.cpp [hunk 13]
```diff
@@ -0,0 +1,689 @@
...
+        expect (chanAmt == XRP (2000));
+
```
Comment (miguelportilla) [inline src/ripple/app/tests/PayChan_test.cpp]: Should probably check alice's balance here, just to verify all is peachy.

File: src/ripple/app/tx/impl/PayChan.cpp [hunk 14]
```diff
@@ -0,0 +1,470 @@
...
+            superior claim.
+        PublicKey
+            The key that will sign claims against the channel.
+        CancelAfter (optional)
+            Any channel transaction that touches this channel after the
+            `CancelAfter` time will close it.
+        DestinationTag (optional)
+            Required if the destination has lsfRequireDestTag set.
+        SourceTag (optional)
+
```
Comment (miguelportilla) [inline src/ripple/app/tx/impl/PayChan.cpp]: Description?

File: src/ripple/app/tx/impl/PayChan.cpp [hunk 15]
```diff
@@ -0,0 +1,470 @@
...
+    view.erase (slep);
+    return tesSUCCESS;
+}
+
+//------------------------------------------------------------------------------
+
+TER
+PayChanCreate::preflight (PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled (featurePayChan, ctx.app.config ().features))
```
Comment (miguelportilla) [inline src/ripple/app/tx/impl/PayChan.cpp]: We seem to be allowing negative values through to sfSettleDelay. Since its unsigned, we might end up with an erroneous value.

File: src/ripple/app/tests/PayChan_test.cpp [hunk 16]
```diff
@@ -0,0 +1,689 @@
...
+    {
+        testSimple ();
+        testCancelAfter ();
+        testSettleDelay ();
+        testExpiration ();
+        testCloseDry ();
+        testDefaultAmount ();
+        testDisallowXRP ();
+        testDstTag ();
+        testMultiple ();
```
Comment (miguelportilla) [inline src/ripple/app/tests/PayChan_test.cpp]: Add test for SrcTag?

File: src/ripple/app/tests/PayChan_test.cpp [hunk 17]
```diff
@@ -0,0 +1,689 @@
...
+    run () override
+    {
+        testSimple ();
+        testCancelAfter ();
+        testSettleDelay ();
+        testExpiration ();
+        testCloseDry ();
+        testDefaultAmount ();
+        testDisallowXRP ();
+        testDstTag ();
```
Comment (miguelportilla) [inline src/ripple/app/tests/PayChan_test.cpp]: Add test for SrcTag?

File: src/ripple/app/tests/PayChan_test.cpp [hunk 18]
```diff
@@ -0,0 +1,689 @@
...
+    }
+
+    static Buffer
+    signClaimAuth (PublicKey const& pk,
+        SecretKey const& sk,
+        uint256 const& channel,
+        STAmount const& authAmt)
+    {
+        Serializer msg;
+        std::uint8_t const hdr[4] = {'C', 'L', 'M', 0};
```
Comment (nbougalis) [inline src/ripple/app/tests/PayChan_test.cpp]: Should this be a `HashPrefix`? See: https://github.com/ripple/rippled/blob/develop/src/ripple/protocol/impl/HashPrefix.cpp

File: src/ripple/app/tests/PayChan_test.cpp [hunk 19]
```diff
@@ -0,0 +1,689 @@
...
+        msg.add64 (authAmt.xrp().drops());
+        return sign (pk, sk, msg.slice ());
+    }
+
+    static
+    STAmount
+    channelBalance (ReadView const& view, uint256 const& chan)
+    {
+        Keylet const k (ltPAYCHAN, chan);
+        auto const slep = view.read (k);
```
Comment (nbougalis) [inline src/ripple/app/tests/PayChan_test.cpp]: `auto const slep = view.read ({ltPAYCHAN, chan});`

File: src/ripple/app/tx/impl/PayChan.cpp [hunk 20]
```diff
@@ -0,0 +1,470 @@
...
+    if (! ctx.rules.enabled(featurePayChan,
+            ctx.app.config().features))
+        return temDISABLED;
+
+    auto const ret = preflight1 (ctx);
+    if (!isTesSuccess (ret))
+        return ret;
+
+    auto const bal = ctx.tx[~sfBalance];
+    auto const amt = ctx.tx[~sfAmount];
```
Comment (miguelportilla) [inline src/ripple/app/tx/impl/PayChan.cpp]: Suggest declaring only before use.

File: src/ripple/app/tx/impl/PayChan.cpp [hunk 21]
```diff
@@ -0,0 +1,470 @@
...
+        auto const page = (*slep)[sfOwnerNode];
+        TER const ter = dirDelete (view, true, page, keylet::ownerDir (src).key,
+            key, false, page == 0, j);
+        if (!isTesSuccess (ter))
+            return ter;
+    }
+
+    // Transfer amount back to owner, decrement owner count
+    auto const sle = view.peek (keylet::account (src));
+    if (!sle)
```
Comment (nbougalis) [inline src/ripple/app/tx/impl/PayChan.cpp]: Under which condition is this possible?

File: src/ripple/app/tx/impl/PayChan.cpp [hunk 22]
```diff
@@ -0,0 +1,470 @@
...
+
+    auto const ret = preflight1 (ctx);
+    if (!isTesSuccess (ret))
+        return ret;
+
+    if (!isXRP (ctx.tx[sfAmount]) || (ctx.tx[sfAmount] <= beast::zero))
+        return temBAD_AMOUNT;
+
+    if (ctx.tx[sfAccount] == ctx.tx[sfDestination])
+        return temMALFORMED;
```
Comment (nbougalis) [inline src/ripple/app/tx/impl/PayChan.cpp]: Use `temDST_IS_SRC` instead.

File: src/ripple/app/tests/PayChan_test.cpp [hunk 23]
```diff
@@ -668,6 +668,68 @@ struct PayChan_test : public beast::unit_test::suite
...
+        }
+        env (create (alice, bob, channelFunds, settleDelay, pk));
+        env.close();
+        auto const chan2Str = to_string (channel (*env.current (), alice, bob));
+        {
+            auto const r =
+                    env.rpc ("account_channels", alice.human (), bob.human ());
+            expect (r[jss::result][jss::channels].size () == 2);
+            expect (chan1Str != chan2Str);
+            for(std::uint32_t i = 0; i<2; ++i)
```
Comment (miguelportilla) [inline src/ripple/app/tests/PayChan_test.cpp]: I'd prefer the following. Especially since chas2Str can be absent and the loop would pass.
expect(r[jss::result][jss::channels][0u][jss::channel_id] == chan2Str ||
            r[jss::result][jss::channels][1u][jss::channel_id] == chan2Str);

File: src/ripple/app/tests/PayChan_test.cpp [hunk 24]
```diff
@@ -0,0 +1,689 @@
...
+        auto const alice = Account ("alice");
+        auto const bob = Account ("bob");
+        env.fund (XRP (10000), alice, bob);
+        auto const pk = alice.pk ();
+        auto const settleDelay = 100s;
+        env (create (alice, bob, XRP (1000), settleDelay, pk));
+        auto const chan = channel (*env.current (), alice, bob);
+        expect (channelBalance (*env.current (), chan) == XRP (0));
+        expect (channelAmount (*env.current (), chan) == XRP (1000));
+
```
Comment (miguelportilla) [inline src/ripple/app/tests/PayChan_test.cpp]: Should probably check alice's balance after create, to verify the amount was sequestered.

File: src/ripple/protocol/impl/TxFormats.cpp [hunk 25]
```diff
@@ -113,6 +113,26 @@ TxFormats::TxFormats ()
...
         << SOElement (sfSignerEntries,       SOE_OPTIONAL)
         ;
+
+    add ("PaymentChannelCreate", ttPAYCHAN_CREATE) <<
+            SOElement (sfDestination,       SOE_REQUIRED) <<
+            SOElement (sfAmount,            SOE_REQUIRED) <<
+            SOElement (sfSettleDelay,       SOE_REQUIRED) <<
+            SOElement (sfPublicKey,         SOE_REQUIRED) <<
+            SOElement (sfCancelAfter,       SOE_OPTIONAL) <<
+            SOElement (sfDestinationTag,    SOE_OPTIONAL);
```
Comment (nbougalis) [inline src/ripple/protocol/impl/TxFormats.cpp]: I'd suggest allowing `sfInvoiceID` as an optional field as well: it's a `uint256` which is _almost_ orthogonal to the destination tag and affords a lot more space for tracking multiple concurrent flows and associating a particular channel with a unique, user-selectable identifier.

File: src/ripple/app/tx/impl/PayChan.cpp [hunk 26]
```diff
@@ -0,0 +1,470 @@
...
+        auto const balance = (*sle)[sfBalance];
+        auto const reserve =
+            ctx_.view ().fees ().accountReserve ((*sle)[sfOwnerCount] + 1);
+
+        if (balance < reserve)
+            return tecINSUFFICIENT_RESERVE;
+
+        if (balance < reserve + ctx_.tx[sfAmount])
+            return tecUNFUNDED;
+    }
```
Comment (nbougalis) [inline src/ripple/app/tx/impl/PayChan.cpp]: I believe that this block, as well as the next (so lines 176-201) can be handled in `preclaim`. Is there a reason why we do this in `doApply`?

File: src/ripple/app/tx/impl/PayChan.cpp [hunk 27]
```diff
@@ -0,0 +1,470 @@
...
+    auto const txAccount = ctx_.tx[sfAccount];
+    auto const expiration = (*slep)[~sfExpiration];
+
+    {
+        auto const cancelAfter = (*slep)[~sfCancelAfter];
+        auto const closeTime =
+            ctx_.view ().info ().parentCloseTime.time_since_epoch ().count ();
+        if ((cancelAfter && closeTime >= *cancelAfter) ||
+            (expiration && closeTime >= *expiration))
+            return closeChannel (
```
Comment (nbougalis) [inline src/ripple/app/tx/impl/PayChan.cpp]: I have concerns: a "fund" operation on an expired channel will return `tesSUCCESS` from here; this makes deciding what happened a nightmare. This is a problem. I suggest we add a new `tecEXPIRED` code and do something like the following:

```
auto ret = closeChannel (
    slep, ctx_.view (), k.key, ctx_.app.journal ("View"));

if (isTesSuccess(ret))
    ret = tecEXPIRED;

return ret;
```

File: src/ripple/app/tx/impl/PayChan.cpp [hunk 28]
```diff
@@ -0,0 +1,470 @@
...
+    if (src != txAccount)
+        // only the owner can add funds or extend
+        return tecNO_PERMISSION;
+
+    if (auto extend = ctx_.tx[~sfExpiration])
+    {
+        auto minExpiration =
+            ctx_.view ().info ().parentCloseTime.time_since_epoch ().count () +
+                (*slep)[sfSettleDelay];
+        if (expiration && *expiration < minExpiration)
```
Comment (nbougalis) [inline src/ripple/app/tx/impl/PayChan.cpp]: Little known-fact: you can skip the `*` here: `boost::optional` transparently overloads operators. See: http://www.boost.org/doc/libs/1_61_0/libs/optional/doc/html/boost_optional/tutorial/relational_operators.html

I'm not suggesting you change anything; just pointing out it's possible. Caveat emptor: I don't know if `std::optional` will have the same behavior.

File: src/ripple/app/tx/impl/PayChan.cpp [hunk 29]
```diff
@@ -0,0 +1,470 @@
...
+    AccountID const txAccount = ctx_.tx[sfAccount];
+
+    auto const curExpiration = (*slep)[~sfExpiration];
+    {
+        auto const cancelAfter = (*slep)[~sfCancelAfter];
+        auto const closeTime =
+            ctx_.view ().info ().parentCloseTime.time_since_epoch ().count ();
+        if ((cancelAfter && closeTime >= *cancelAfter) ||
+            (curExpiration && closeTime >= *curExpiration))
+            return closeChannel (
```
Comment (nbougalis) [inline src/ripple/app/tx/impl/PayChan.cpp]: Same concern as the `fund` case. Returning `tesSUCCESS` if I try to claim on an expired channels complicates things. I'd prefer `tecEXPIRED`.

File: src/ripple/app/tx/impl/PayChan.cpp [hunk 30]
```diff
@@ -0,0 +1,470 @@
...
+        {
+            // Check the signature
+            // The signature isn't needed if txAccount == src, but if it's
+            // present, check it
+
+            PublicKey const pk ((*slep)[sfPublicKey]);
+            if (ctx_.tx[sfPublicKey] != pk)
+                return temBAD_SIGNER;
+            Serializer msg;
+            std::uint8_t const hdr[4] = {'C', 'L', 'M', 0};
```
Comment (nbougalis) [inline src/ripple/app/tx/impl/PayChan.cpp]: We duplicate this bit of code in the test, as part of `signClaimAuth`. I would suggest having a function that does this once:

```
void serializePayChanMessage (
    Serializer& msg,
    uint256 const& key,
    XRPAmount const& amount)
{
    std::uint8_t const hdr[4] = {'C', 'L', 'M', 0};
    msg.addRaw (hdr, 4);
    msg.add256 (key);
    msg.add64 (amount.drops());
}
```

If you don't want to pass the `Serializer` and are ok with a memory allocation and a copy, you can do:

```
Buffer serializePayChanMessage (
    uint256 const& key,
    XRPAmount const& amount)
{
    std::uint8_t const hdr[4] = {'C', 'L', 'M', 0};

    Serializer msg;
    msg.addRaw (hdr, 4);
    msg.add256 (key);
    msg.add64 (amount.drops());

    return { msg.slice() };
}
```

File: src/ripple/app/tx/impl/PayChan.cpp [hunk 31]
```diff
@@ -0,0 +1,470 @@
...
+            msg.addRaw (hdr, 4);
+            msg.add256 (k.key);
+            msg.add64 (authAmt.drops ());
+            if (!verify (pk, msg.slice (), *sig, /*canonical*/ true))
```
Comment (nbougalis) [inline src/ripple/app/tx/impl/PayChan.cpp]: The `true` isn't strictly necessary - it's defaulted and unlikely to ever change, so you could skip it. On the other hand, we want to enforce canonical signatures, so "reinforcing" that by explicitly demanding canonical verification isn't a bad idea.

I'm not feeling strongly about this either way.

File: src/ripple/app/tx/impl/PayChan.cpp [hunk 32]
```diff
@@ -0,0 +1,477 @@
...
+        return temBAD_AMOUNT;
+
+    if (bal && amt && *bal > *amt)
+        return tecNO_PERMISSION;
+
+    auto const flags = ctx.tx.getFlags ();
+    if ((flags & tfClose) && (flags & tfRenew))
+        return temMALFORMED;
+
+    if (auto const& sig = ctx.tx[~sfSignature])
```
Comment (nbougalis) [inline src/ripple/app/tx/impl/PayChan.cpp]: Why `&` here?

File: src/ripple/app/tx/impl/PayChan.cpp (hunk 29 above)
Comment (JoelKatz) [inline src/ripple/app/tx/impl/PayChan.cpp]: We have no choice but to return `tesSUCCESS` if we do anything other than claim a fee. That's the nature of the design at this point.

File: src/ripple/app/tx/impl/PayChan.cpp (hunk 27 above)
Comment (nbougalis) [inline src/ripple/app/tx/impl/PayChan.cpp]: Addendum: `tesSUCCESS` is the best we can do now because `closeChannel` will transfer any unclaimed funds back to the source,and a `tec` code can't really perform transactions.

The documentation needs to explicitly explain that "fund" really means "fund if not expired, do nothing otherwise" and that a `tesSUCCESS` may mean "funding succeeded!" or "channel expired."

File: src/ripple/app/tx/impl/PayChan.cpp (hunk 29 above)
Comment (nbougalis) [inline src/ripple/app/tx/impl/PayChan.cpp]: Yep. I'm convinced.

File: src/ripple/app/tx/impl/PayChan.cpp (hunk 15 above)
Comment (seelabs) [inline src/ripple/app/tx/impl/PayChan.cpp]: By the time we get to preflight, it's too late to do this check. I think we'd need a check when the json is converted to an STObject. If we want to do this it sounds like this should be a general check and not something specific to payment channels. Does anyone know if we already have these checks in place? Do we want to add them?

File: src/ripple/app/tests/PayChan_test.cpp (hunk 17 above)
Comment (seelabs) [inline src/ripple/app/tests/PayChan_test.cpp]: There are situations where the dst tag is required, and we test for those. There are no analogous tests for src tag.

File: src/ripple/app/tx/impl/PayChan.cpp (hunk 28 above)
Comment (seelabs) [inline src/ripple/app/tx/impl/PayChan.cpp]: Yeah, I knew about that (I know because Scott Schurr used it in some code I reviewed). I prefer the non-magic version.

File: src/ripple/app/tx/impl/PayChan.cpp (hunk 21 above)
Comment (seelabs) [inline src/ripple/app/tx/impl/PayChan.cpp]: It can't happen. I'll remove the check.

File: src/ripple/app/tx/impl/PayChan.cpp (hunk 32 above)
Comment (seelabs) [inline src/ripple/app/tx/impl/PayChan.cpp]: No reason really. I changed it to a value (<pedantic> The ref doesn't _hurt_ us here, but given the two equivalent ways I agree a value type here is expected).

**Comment by seelabs** [comment]
@nbougalis @miguelportilla I pushed two new commits that should address all the review feedback except two:

1) Mickey's suggestion to check for negative settle times
2) Nik's suggestion to add a sfInvoiceID field

File: src/ripple/app/tx/impl/PayChan.cpp (hunk 15 above)
Comment (seelabs) [inline src/ripple/app/tx/impl/PayChan.cpp]: @miguelportilla There appears to be a bug in `range_check_cast`. I opened a separate PR to address the issue.

File: src/ripple/protocol/impl/TxFormats.cpp (hunk 25 above)
Comment (seelabs) [inline src/ripple/protocol/impl/TxFormats.cpp]: @JoelKatz What do you think about this? Should I add the invoice field?

**Comment by seelabs** [comment]
Rebased onto 0.33.0-b1

**Comment by seelabs** [comment]
Rebased onto 0.33.0-b2

**Comment by seelabs** [comment]
@nbougalis @miguelportilla Added a sfInvoiceID field (as suggested by Nik - David thought it was a good idea as well). I believe that addresses all the feedback.

**Comment by seelabs** [comment]
Pushed a commit to add an InvoiceID to the claim tx. If the channel is created with a InvoiceID, the claim tx must must have the matching InvoiceID. If it does not `temBAD_INVOICE` is returned.

In addition to a code review, please think about if the interface feels right. Let's get the interface right or add invoices later if we can't get it right now.

File: src/ripple/app/tx/impl/PayChan.cpp [hunk 33]
```diff
@@ -358,7 +374,7 @@ PayChanClaim::preflight (PreflightContext const& ctx)
 
         auto const reqBalance = bal->xrp ();
         auto const authAmt = [&] {
-            if (auto const& amt = ctx.tx[~sfAmount])
+            if (auto const amt = ctx.tx[~sfAmount])
```
Comment (miguelportilla) [inline src/ripple/app/tx/impl/PayChan.cpp]: amt was already declared on 355 and you're auto capturing by ref.

File: src/ripple/app/tests/PayChan_test.cpp [hunk 34]
```diff
@@ -106,7 +106,8 @@ struct PayChan_test : public beast::unit_test::suite
         NetClock::duration const& settleDelay,
         PublicKey const& pk,
         boost::optional<NetClock::time_point> cancelAfter = boost::none,
-        boost::optional<std::uint32_t> dstTag = boost::none)
+        boost::optional<std::uint32_t> dstTag = boost::none,
+        boost::optional<uint256> invoiceID = boost::none)
```
Comment (miguelportilla) [inline src/ripple/app/tests/PayChan_test.cpp]: Super nit pick: These can both be const refs.

File: src/ripple/app/tx/impl/PayChan.cpp [hunk 35]
```diff
@@ -0,0 +1,512 @@
...
+    {
+        if (!(ctx.tx[~sfPublicKey] && bal))
+            return temMALFORMED;
+
+        // Check the signature
+        // The signature isn't needed if txAccount == src, but if it's
+        // present, check it
+
+        auto const reqBalance = bal->xrp ();
+        auto const authAmt = [&] {
```
Comment (nbougalis) [inline src/ripple/app/tx/impl/PayChan.cpp]: Why not:

    auto const authAmt = (amt ? amt->xrp() : reqBalance);

File: src/ripple/app/tests/PayChan_test.cpp (hunk 34 above)
Comment (seelabs) [inline src/ripple/app/tests/PayChan_test.cpp]: Already fixed in a later commit.

File: src/ripple/app/tests/PayChan_test.cpp [hunk 36]
```diff
@@ -779,6 +782,64 @@ struct PayChan_test : public beast::unit_test::suite
...
+            auto const r =
+                env.rpc ("account_channels", alice.human (), carol.human ());
+            BEAST_EXPECT (r[jss::result][jss::channels].size () == 1);
+            BEAST_EXPECT (r[jss::result][jss::channels][0u][jss::channel_id] == chan);
+            BEAST_EXPECT (r[jss::result][jss::channels][0u][jss::destination_tag] == dstTag);
+            BEAST_EXPECT (!r[jss::result][jss::channels][0u].isMember(jss::invoice_id));
+        }
+        {
+            Blob raw(256/8);
+            std::iota(raw.begin(), raw.end(), 0);
```
Comment (miguelportilla) [inline src/ripple/app/tests/PayChan_test.cpp]: #include <numeric>

**Comment by seelabs** [comment]
@nbougalis @miguelportilla I pushed a commit that removes the InvoiceID additions and leaves some minor cleanups.

**Comment by miguelportilla** [comment]
:+1:

**Comment by seelabs** [comment]
Squashed and marking as passed (I spoke with @nbougalis ; He gave a thumbs-up in a DM).
